heron/metricsmgr/src/java/com/twitter/heron/metricsmgr/sink/metricscache/MetricsCacheSink.java

```diff
@@ -0,0 +1,417 @@
+
+            // Update Metrics
+            sinkContext.exportCountMetric(TMASTER_LOCATION_UPDATE_COUNT, 1);
+          }
+        }
+
+        // Schedule itself in future
+        tMasterLocationStarter.schedule(this, checkIntervalSec, TimeUnit.SECONDS);
+      }
+    };
+
+    // First Entry
+    tMasterLocationStarter.schedule(runnable, checkIntervalSec, TimeUnit.SECONDS);
+  }
+
+  @Override
+  public void processRecord(MetricsRecord record) {
+    LOG.info("metricscache sink processRecord");
+    // Format it into TopologyMaster.PublishMetrics
+
+    // The format of source is "host:port/componentName/instanceId"
+    // So source.split("/") would be an array with 3 elements:
+    // ["host:port", componentName, instanceId]
+    String[] sources = record.getSource().split("/");
+    String hostPort = sources[0];
+    String componentName = sources[1];
+    String instanceId = sources[2];
+
+    TopologyMaster.PublishMetrics.Builder publishMetrics =
+        TopologyMaster.PublishMetrics.newBuilder();
+
+    for (MetricsInfo metricsInfo : tMasterMetricsFilter.filter(record.getMetrics())) {
+      // We would filter out unneeded metrics
+      TopologyMaster.MetricDatum metricDatum = TopologyMaster.MetricDatum.newBuilder().
+          setComponentName(componentName).setInstanceId(instanceId).setName(metricsInfo.getName()).
+          setValue(metricsInfo.getValue()).setTimestamp(record.getTimestamp()).build();
+      publishMetrics.addMetrics(metricDatum);
+    }
+
+    for (ExceptionInfo exceptionInfo : record.getExceptions()) {
+      TopologyMaster.TmasterExceptionLog exceptionLog =
+          TopologyMaster.TmasterExceptionLog.newBuilder()
+              .setComponentName(componentName)
+              .setHostname(hostPort)
+              .setInstanceId(instanceId)
+              .setStacktrace(exceptionInfo.getStackTrace())
+              .setLasttime(exceptionInfo.getLastTime())
+              .setFirsttime(exceptionInfo.getFirstTime())
+              .setCount(exceptionInfo.getCount())
+              .setLogging(exceptionInfo.getLogging()).build();
+      publishMetrics.addExceptions(exceptionLog);
+    }
+
+    metricsCommunicator.offer(publishMetrics.build());
+
+    // Update metrics
+    sinkContext.exportCountMetric(RECORD_PROCESS_COUNT, 1);
+    sinkContext.exportCountMetric(METRICS_COUNT, publishMetrics.getMetricsCount());
+    sinkContext.exportCountMetric(EXCEPTIONS_COUNT, publishMetrics.getExceptionsCount());
+  }
+
+  @Override
+  public void flush() {
+    // We do nothing here but update metrics
+    sinkContext.exportCountMetric(TMASTER_RESTART_COUNT,
+        metricsCacheClientService.startedAttempts.longValue());
+  }
+
+  @Override
+  public void close() {
+    metricsCacheClientService.close();
+    metricsCommunicator.clear();
+  }
+
+  /////////////////////////////////////////////////////////
+  // Following protected methods should be used only for unit testing
```

Comment:
Please remove and instead annotate methods @VisibleForTesting. Can they be package private instead of protected?